From 4eea1370f89ba14fff2b3150e40f3908f8971193 Mon Sep 17 00:00:00 2001
From: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
Date: Tue, 30 Sep 2014 17:26:44 +0100
Subject: [PATCH 329/351] switch: arizona: Additional settings to improve
 accuracy of HP detect

If the TST_CAP_SEL bits aren't set correctly there will be a 100k load
along side the headphones, which will affect the accurary towards the
very top of the detection range.

Change-Id: I354b2f5c0eed83a05ec81c784a20dbf6a2f64ca4
Signed-off-by: Charles Keepax <ckeepax@opensource.wolfsonmicro.com>
---
 drivers/switch/switch-arizona.c |   17 +++++++++++++++--
 1 file changed, 15 insertions(+), 2 deletions(-)

diff --git a/drivers/switch/switch-arizona.c b/drivers/switch/switch-arizona.c
index 2ec2cc0..02f9710 100644
--- a/drivers/switch/switch-arizona.c
+++ b/drivers/switch/switch-arizona.c
@@ -244,6 +244,7 @@ static void arizona_extcon_hp_clamp(struct arizona_extcon_info *info,
 {
 	struct arizona *arizona = info->arizona;
 	unsigned int mask, val = 0;
+	unsigned int cap_sel = 0;
 	int ret;
 
 	switch (arizona->type) {
@@ -255,10 +256,22 @@ static void arizona_extcon_hp_clamp(struct arizona_extcon_info *info,
 	case WM5110:
 		mask = ARIZONA_HP1L_SHRTO | ARIZONA_HP1L_FLWR |
 		       ARIZONA_HP1L_SHRTI;
-		if (clamp)
+		if (clamp) {
 			val = ARIZONA_HP1L_SHRTO;
-		else
+			cap_sel = 1;
+		} else {
 			val = ARIZONA_HP1L_FLWR | ARIZONA_HP1L_SHRTI;
+			cap_sel = 3;
+		}
+
+		ret = regmap_update_bits(arizona->regmap,
+					 ARIZONA_HP_TEST_CTRL_1,
+					 ARIZONA_HP1_TST_CAP_SEL_MASK,
+					 cap_sel);
+		if (ret != 0)
+			dev_warn(arizona->dev,
+				"Failed to set TST_CAP_SEL: %d\n",
+				 ret);
 		break;
 	default:
 		mask = ARIZONA_RMV_SHRT_HP1L;
-- 
1.7.9.5

